### PR TITLE
Particle sampler implementation

### PR DESCRIPTION
**Pull request by denainjs**
Library for particle samplers in Turing.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+[deps]
+AbstractMCMC = "80f14c24-f653-4e6a-9b94-39d6b0f70001"
+Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
+DynamicPPL = "366bfd00-2699-11ea-058f-f148b4cae6d8"
+Libtask = "6f1fad26-d15e-5dc8-ae53-837a1d7b8c9f"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+StatsFuns = "4c63d2b9-4356-54db-8cca-17b64c39e42c"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Turing = "fce5fe82-541a-59a6-adf8-730c64b5f9a0"
```
Comment (devmotion) [inline Project.toml:14]: The Manifest.toml file should be removed and Project.toml should be cleaned: test dependencies have to be moved to a separate section (e.g., this package shouldn't depend on Test or Turing, and probably also not on DynamicPPL), and compatibility bounds have to be specified.

File: README.md [hunk 2]
```diff
@@ -1,23 +1 @@
...
-5. Pitt, Michael K., and Neil Shephard. "Filtering via simulation: Auxiliary particle filters." Journal of the American statistical association 94, no. 446 (1999): 590-599.
-
-6. Doucet, Arnaud, Nando de Freitas, and Neil Gordon. "Sequential Monte Carlo Methods in Practice."
-
-7. Del Moral, Pierre, Arnaud Doucet, and Ajay Jasra. "Sequential Monte Carlo samplers." Journal of the Royal Statistical Society: Series B (Statistical Methodology) 68, no. 3 (2006): 411-436.
-
-
-
-
+# AdvancedPS.jl
```
Comment (devmotion) [inline README.md:1]: Why are the references removed?

File: src/particlecontainer.jl [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    vals = eltype(pc.vals)[fork(p) for p in pc.vals]
+
+    # copy weights
+    logWs = copy(pc.logWs)
+
+    ParticleContainer(vals, logWs)
+end
+
+"""
+    reset_logweights!(pc::ParticleContainer)
```
Comment (devmotion) [inline src/particlecontainer.jl:43]: Copying messed up the docstrings.

File: test/resampling.jl [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    resSystematic = resample_systematic(D, num_samples )
+    resStratified = resample_stratified(D, num_samples )
+    resMultinomial= resample_multinomial(D, num_samples )
+    resResidual   = resample_residual(D, num_samples )
+
+
+    @test sum(resSystematic .== 2) ≈ (num_samples * 0.4) atol=1e-3*num_samples
+    @test sum(resStratified .== 2) ≈ (num_samples * 0.4) atol=1e-3*num_samples
+    @test sum(resMultinomial .== 2) ≈ (num_samples * 0.4) atol=1e-2*num_samples
+    @test sum(resResidual .== 2) ≈ (num_samples * 0.4) atol=1e-2*num_samples
```
Comment (devmotion) [inline test/resampling.jl:16]:
```suggestion
    @test count(==(2), resSystematic) ≈ 0.4 * num_samples 0.4atol=1e-3*num_samples
    @test count(==(2), resStratified) ≈ 0.4 * num_samples atol=1e-3*num_samples
    @test count(==(2), resMultinomial) ≈ 0.4 * num_samples atol=1e-2*num_samples
    @test count(==(2), resResidual) ≈ 0.4 * num_samples atol=1e-2*num_samples
```
Why are the different tolerances needed BTW?

File: Project.toml (hunk 1 above)
Comment (denainjs) [reply Project.toml]: what do you mean by 'moved to a separated section'?

File: README.md (hunk 2 above)
Comment (denainjs) [reply README.md:1]: my bad, changed

File: Project.toml (hunk 1 above)
Comment (devmotion) [reply Project.toml]: Yes, exactly :+1:

**Comment by cpfiffer** [comment]
I share @devmotion's concern that it's hard to review this. Could we basically just copy all the Turing code first in a separate PR, merge that through first, and then use the new master to compare the changes made as part of this PR?

**Comment by denainjs** [comment]
AdvancedPS here is supposed to work with the Turing samplers introduced in [this PR](https://github.com/TuringLang/Turing.jl/pull/1381).
I don't think it addresses the issues in [TuringLang/Turing.jl#1345](TuringLang/Turing.jl#1345).

**Comment by cpfiffer** [comment]
Yeah, that's fine -- it's just difficult to examine the changes because all the code here is "new" to AdvancedPS.jl, so we can't review changes that might be a bad idea because there's nothing to compare to.

**Comment by yebai** [comment]
Close in favour of #14

**Pull request closed without merging**